lib/theme_check/language_server/variable_lookup_finder.rb

```diff
@@ -0,0 +1,263 @@
+# frozen_string_literal: true
+
+module ThemeCheck
+  module LanguageServer
+    module VariableLookupFinder
+      extend self
+
+      UNCLOSED_SQUARE_BRACKET = /\[[^\]]*\Z/
+      UNCLOSED_SINGLE_QUOTE = /'[^']*\Z/
+      UNCLOSED_DOUBLE_QUOTE = /"[^"]*\Z/
+
+      def lookup(content, cursor)
+        potential_lookup = lookup_liquid_variable(content, cursor) || lookup_liquid_tag(content, cursor)
+
+        # And we only return it if it's parsed by Liquid as VariableLookup
+        return unless potential_lookup.is_a?(Liquid::VariableLookup)
+        potential_lookup
+      end
+
+      private
+
+      def cursor_is_on_liquid_variable_lookup_position(content, cursor)
+        previous_char = content[cursor - 1]
+        is_liquid_variable = content =~ Liquid::VariableStart
+        is_in_variable_segment = previous_char =~ /[a-z0-9_.'"-]/i
+        is_on_blank_variable_lookup_position = content[0..cursor - 1] =~ /[{:,]\s+$/
+        (
+          is_liquid_variable && (
+            is_in_variable_segment ||
+            is_on_blank_variable_lookup_position
+          )
+        )
+      end
+
+      def lookup_liquid_variable(content, cursor)
+        return unless cursor_is_on_liquid_variable_lookup_position(content, cursor)
+        start_index = 2
+        end_index = cursor - 1
+
+        # We take the following content
+        # - start after the first two {{
+        # - end at cursor position
+        #
+        # That way, we'll have a partial liquid variable that
+        # can be parsed such that the "last" variable_lookup
+        # will be the one we're trying to complete.
+        markup = content[start_index..end_index]
+
+        # Early return for incomplete variables
+        return empty_lookup if markup =~ /\s+$/
+        return if markup =~ /((?<!\[)['"]|\]|\[)$/
```

Comment:
Should name that regex. That for square brackets thing. Should also make sure that this also happens in liquid tags.